Solutions/Marain.UserNotifications.Client/Menes/Hal/WebLink.cs

```diff
@@ -69,7 +69,7 @@ public WebLink(in JsonElement source)
         /// Either a URI [RFC3986] or URI Template [RFC6570] of the target
         /// resource.
         /// </remarks>
-        public string Href => this.source.GetProperty(HrefProperty).GetString();
+        public string Href => this.source.GetProperty(HrefProperty).GetString() ?? string.Empty;
```

Comment:
I think this might not be the right way to handle this.

I was a bit perplexed as to why this wasn't already an issue: the code was already nullable-aware, so why wasn't this producing a warning before? And it turns out that `JsonElement.GetString` has changed. Its [.NET 6.0 definition](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonelement.getstring?view=net-6.0) returns a `string?` but its [.NET Core 3.1 version](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonelement.getstring?view=netcore-3.1) seems to return `string`; I'm guessing it was null-oblivious at that point, although that seems a little odd given that as far as I can remember, `JsonElement` and `JsonProperty` in `System.Text.Json` were brand new APIs introduced to .NET Core 3.0 at exactly the same time nullable reference types were introduced so it's weird it wasn't null-aware from the start.

Anyway, history aside, the question is: what is the right behaviour here? The `Href` property in this class is distinctive in that it provides a non-nullable `string`, where most of the rest are nullable. This reflects the fact that in a `WebLink`, a hyperlink is non-optional: if something doesn't have a hyperlink (i.e., an href) then it's not a link.

So the problem with this modification is that it just shuts the compiler up by sweeping the problem under the carpet. The compiler has helpfully revealed a problem but the change in this PR essentially throws that information away. We can do better.

First, let's look at the situation in which this might return null. It will happen with the following JSON:

```json
{"href": null }
```

That's definitely not a valid web link. I can see three options here:

1. check that the href is present in the constructor and throw an `ArgumentException` if it's not—this means `WebLink` would then be correct by construction: it would be impossible to create one in cases where the href is absent
2. detect the absence of the href attribute in the `Href` property, and throw an `InvalidOperationException` in cases where someone tries to retrieve the `Href`
3. pretend everything is fine by returning an empty string instead of null

I initially thought we had a potential problem with 1: this is a `struct` so our constructor won't necessarily run. For example, if someone defines a class with a field of type `WebLink`, then unless they explicitly initialize it, that `WebLink` will get the default initialization, with the result that its `source` field will be a `JsonElement` that doesn't refer to anything. However, the effect of this is that any attempt to use that `JsonElement` throws an `InvalidOperationException`. Since every member of `WebLink` uses that `JsonElement` it will throw that exception if anyone attempts this. That's the exception I would have thrown if we explicitly detected this case; the only thing we could conceivably improve with additional handling is to throw the same exception but with a different error message, explaining that you need to construct a `WebLink` explicitly. But I'm not that fussed about that.

The one problem with 1 is that there may be scenarios where you want to be able to represent invalid data. E.g., if an invalid request has been supplied and we want to produce helpful error messages describing what's wrong, that can be made harder when an object model simply refuses to be constructed because one little bit is wrong.

That said, I'm not sure 2 works any better. If this "parse an invalid input so we can explain exactly how it's wrong" scenario is one we care about, we might need to do more. Really, we'd want some sort of validation process, and since I know such a thing is in the works, it might be better not to attempt to solve that problem in this type at all.

I think 3 (which is what we have in this PR) is definitely not the right thing to do, because it just passes the buck. At best this will cause exceptions slightly later on in processing, but really we should just report that we've detected a problem.

So I think the best bet here is to add code in the constructor that verifies that the href is present and is a string (probably needs to be a non-empty one: I don't think empty strings are ever valid here), and then here we can just use the `!` operator in this property because we already checked during construction that a null result can't occur here.

However, I'm seeking input from colleagues who may have a different view.

Reply:
This is definitely a problem with the hand-crafted prototype code in use in this client.

Its deficiencies informed the model for [Corvus.JsonSchema](https://github.com/corvus-dotnet/Corvus.JsonSchema/). If you look at the behaviour there, an instance of a Json 'type' can be freely constructed from any other type (or JsonElement). In that sense invalid input is always parsed "correctly" (i.e. you just "say" that it is the type you want it to be). However, there is a separate _validate_ step you use to ensure that it conforms to the schema that underlies the type, and can therefore be used _as such a type_. A validated duck-typing model leveraging the fact that the types are generated from JsonSchema.

There are then property getters which return concrete types for JSON primitives e.g. `JsonString`. The 'default' values for those properties are structurally equivalent  (e.g. for comparison, casting purposes) to `JsonNull` so can be returned safely. There are also `TryGetProperty()` members of the usual TryGet pattern.

The problem in the protype code is that it is not defensive against the fact that it may not be valid, and has lost any connection with underlying schema and validation - because, of course, it never had it in the first place. It was validating alloc-free mechanisms for accessing underlying System.Text.Json representations of the dafta.

With all that preamble in mind, I would suggest that your solution is possibly the right one.

Another one to consider would be to define schema for these types, with appropriate validation, and generate them using the new Corvus.JsonSchema tooling?

I think the schema [here](https://github.com/menes-dotnet/Menes/blob/feature/typegenerator/Solutions/Menes/ResourceSample/PeopleApi.json) might be useful in the light of that.

Reply:
If you want to use the tool to experimentally generate relevant types:

`--rootNamespace Corvus.PeopleApi --useSchema Draft202012 --rootPath #/components/schemas/PersonListResource --outputRootTypeName PersonListResource --outputMapFile PersonListResource.map --outputPath [PathToMenes]\\Menes\\Solutions\\Menes\\ResourceSample [PathToMenes]\\Menes\\Solutions\\Menes\\ResourceSample\\PeopleApi.json`

Reply:
And here is a generated set of sample files based on PersonListResource to illustrate what you get out at the other end
[GeneratedExample.zip](https://github.com/marain-dotnet/Marain.UserNotifications/files/8131552/GeneratedExample.zip)
.